File: backend/src/main/java/com/orderflow/backend/service/ProductService.java
package com.orderflow.backend.service;

import com.orderflow.backend.model.Product;
import com.orderflow.backend.repository.ProductRepository;
import org.springframework.stereotype.Service;
import com.orderflow.backend.exception.ResourceNotFoundException;
import java.util.List;

@Service
public class ProductService {

    private final ProductRepository productRepository;

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    // Create product
    public Product createProduct(Product product) {
        return productRepository.save(product);
    }

    // Get all products
    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    // Get product by ID
  public Product getProductById(Long id) {
    return productRepository.findById(id)
            .orElseThrow(() ->
                    new ResourceNotFoundException(
                            "Product not found with id: " + id
                    ));
}

    // Search products by name
    public List<Product> searchProducts(String name) {
        return productRepository.findByNameContainingIgnoreCase(name);
    }

    // Get products by category
    public List<Product> getProductsByCategory(String category) {
        return productRepository.findByCategoryIgnoreCase(category);
    }

    // Get active products
    public List<Product> getActiveProducts() {
        return productRepository.findByActiveTrue();
    }

    // Update product
    public Product updateProduct(Long id, Product updatedProduct) {

        Product existingProduct = getProductById(id);

        existingProduct.setName(updatedProduct.getName());
        existingProduct.setDescription(updatedProduct.getDescription());
        existingProduct.setPrice(updatedProduct.getPrice());
        existingProduct.setStockQuantity(updatedProduct.getStockQuantity());
        existingProduct.setCategory(updatedProduct.getCategory());
        existingProduct.setActive(updatedProduct.getActive());

        return productRepository.save(existingProduct);
    }

    // Delete product
    public void deleteProduct(Long id) {

        Product existingProduct = getProductById(id);

        productRepository.delete(existingProduct);
    }
}

File: frontend/src/services/productService.ts
import api from "./api";

import type {
  Product,
} from "../types/product";

export interface ProductRequest {
  name: string;
  description: string;
  price: number;
  stockQuantity: number;
  category: string;
  active: boolean;
}

export const getAllProducts =
  async (): Promise<Product[]> => {

    const response =
      await api.get<Product[]>(
        "/products"
      );

    return response.data;
  };

export const createProduct =
  async (
    product: ProductRequest
  ): Promise<Product> => {

    const response =
      await api.post<Product>(
        "/products",
        product
      );

    return response.data;
  };

export const updateProduct =
  async (
    productId: number,
    product: ProductRequest
  ): Promise<Product> => {

    const response =
      await api.put<Product>(
        `/products/${productId}`,
        product
      );

    return response.data;
  };

export const deleteProduct =
  async (
    productId: number
  ): Promise<void> => {

    await api.delete(
      `/products/${productId}`
    );
  };